--- a/whiptcl.c
+++ b/whiptcl.c
@@ -1,4 +1,5 @@
 #include "config.h"
+#include <stdio.h>
 #include <string.h>
 #include <stdlib.h>
 
@@ -138,47 +139,46 @@ static int wtCmd(ClientData clientData, Tcl_Interp * interp, int argc,
     }
     
     if (arg < -1) {
-	/* this could buffer oveflow, bug we're not setuid so I don't care */
-	interp->result = malloc(200);
-	interp->freeProc = TCL_DYNAMIC;
-	sprintf(interp->result, "%s: %s\n", 
+	char *err = Tcl_Alloc(200);
+	snprintf(err, 200, "%s: %s\n",
 		poptBadOption(optCon, POPT_BADOPTION_NOALIAS), 
 		poptStrerror(arg));
+	Tcl_SetResult(interp, err, TCL_DYNAMIC);
 
 	return TCL_ERROR;
     }
 
     if (mode == MODE_NONE) {
-	interp->result = "no dialog mode was specified";
+	Tcl_SetResult(interp, "no dialog mode was specified", TCL_STATIC);
 	return TCL_ERROR;
     } else if (rc) {
-	interp->result = "multiple modes were specified";
+	Tcl_SetResult(interp, "multiple modes were specified", TCL_STATIC);
 	return TCL_ERROR;
     }
 
     if (!(text = poptGetArg(optCon))) {
-	interp->result = "missing text parameter";
+	Tcl_SetResult(interp, "missing text parameter", TCL_STATIC);
 	return TCL_ERROR;
     }
 
     if (!(nextArg = poptGetArg(optCon))) {
-	interp->result = "height missing";
+	Tcl_SetResult(interp, "height missing", TCL_STATIC);
 	return TCL_ERROR;
     }
     height = strtoul(nextArg, &end, 10);
     if (*end) {
-	interp->result = "height is not a number";
+	Tcl_SetResult(interp, "height is not a number", TCL_STATIC);
 	return TCL_ERROR;
     }
 
     if (!(nextArg = poptGetArg(optCon))) {
-	interp->result = "width missing";
+	Tcl_SetResult(interp, "width missing", TCL_STATIC);
 	return TCL_ERROR;
     }
     width = strtoul(nextArg, &end, 10);
     if (*end) {
-	interp->result = "width is not a number";
+	Tcl_SetResult(interp, "width is not a number", TCL_STATIC);
 	return TCL_ERROR;
     }
 
@@ -187,12 +187,12 @@ static int wtCmd(ClientData clientData, Tcl_Interp * interp, int argc,
       case MODE_RADIOLIST:
       case MODE_CHECKLIST:
 	if (!(nextArg = poptGetArg(optCon))) {
-	    interp->result = "list-height missing";
+	    Tcl_SetResult(interp, "list-height missing", TCL_STATIC);
 	    return TCL_ERROR;
 	}
 	listHeight = strtoul(nextArg, &end, 10);
 	if (*end) {
-	    interp->result = "list-height is not a number";
+	    Tcl_SetResult(interp, "list-height is not a number", TCL_STATIC);
 	    return TCL_ERROR;
 	}
 	break;
@@ -213,35 +213,35 @@ static int wtCmd(ClientData clientData, Tcl_Interp * interp, int argc,
       case MODE_YESNO:
 	rc = messageBox(text, height, width, MSGBOX_YESNO, flags);
 	if (rc == DLG_OKAY)
-	    interp->result = "yes";
+	    Tcl_SetResult(interp, "yes", TCL_STATIC);
 	else 
-	    interp->result = "no";
+	    Tcl_SetResult(interp, "no", TCL_STATIC);
 	if (rc == DLG_ERROR) rc = 0;
 	break;
 
       case MODE_INPUTBOX:
 	rc = inputBox(text, height, width, optCon, flags, &result);
 	if (rc ==DLG_OKAY) {
-	    interp->result = result;
-	    interp->freeProc = TCL_DYNAMIC;
+	    Tcl_SetResult(interp, result, TCL_VOLATILE);
+	    free(result);
 	}
 	break;
 
       case MODE_MENU:
 	rc = listBox(text, height, width, listHeight, optCon, flags, default_item,
 		     &result);
 	if (rc==DLG_OKAY) {
-	    interp->result = result;
-	    interp->freeProc = TCL_DYNAMIC;
+	    Tcl_SetResult(interp, result, TCL_VOLATILE);
+	    free(result);
 	}
 	break;
 
       case MODE_RADIOLIST:
 	rc = checkList(text, height, width, listHeight, optCon, 1, flags,
 		       &selections);
 	if (rc==DLG_OKAY) {
-	    interp->result = selections[0];
-	    interp->freeProc = TCL_DYNAMIC;
+	    Tcl_SetResult(interp, selections[0], TCL_VOLATILE);
 
+	    free(selections[0]);
 	    free(selections);
 	}
 	break;
@@ -251,8 +251,10 @@ static int wtCmd(ClientData clientData, Tcl_Interp * interp, int argc,
 		       &selections);
 
 	if (rc==DLG_OKAY) {
-	    for (next = selections; *next; next++) 
+	    for (next = selections; *next; next++) {
 		Tcl_AppendElement(interp, *next);
+		free(*next);
+	    }
 
 	    free(selections);
 	}
@@ -267,7 +269,7 @@ static int wtCmd(ClientData clientData, Tcl_Interp * interp, int argc,
     newtPopWindow();
 
     if (rc == DLG_ERROR) {
-	interp->result = "bad paramter for whiptcl dialog box";
+	Tcl_SetResult(interp, "bad paramter for whiptcl dialog box", TCL_STATIC);
 	return TCL_ERROR;
     } 
 
@@ -313,7 +315,7 @@ static char * setFullButtons(ClientData data, Tcl_Interp * interp,
     int state;
     
     if ((rc = Tcl_ExprBoolean(interp, val, &state))) {
-	Tcl_FreeResult(interp);
+	Tcl_ResetResult(interp);
 	return "whiptcl_fullbuttons may only contain a boolean value";
     }
 
